### ISSUE TITLE
Remove MultiTenancyStrategy as HHH 6 and Jakarta are in the main

### ISSUE THREAD

**Issue by rsvoboda** [quarkusio/quarkus#32787]
### Description

Remove MultiTenancyStrategy as HHH 6 and Jakarta are in the main

https://github.com/quarkusio/quarkus/blob/main/extensions/hibernate-orm/runtime/src/main/java/io/quarkus/hibernate/orm/runtime/migration/MultiTenancyStrategy.java

` * @deprecated This class should be removed after we're done migrating to Jakarta APIs and Hibernate ORM v6.`

Introduced in https://github.com/quarkusio/quarkus/pull/25796

CC @Sanne

**Comment by yrodiere** [comment]
@Sanne I just had a look at this, and I think we kind of need this strategy in Quarkus itself, for multiple reasons:

1. Our default implementation of `TenantConnectionResolver` behaves differently when using schema multitenancy vs. database multitenancy.
2. We need to determine whether Quarkus needs to pass a `MultiTenantConnectionProvider` to Hibernate ORM. This would be the case only with database/schema multitenancy, and since we have a default implementation of `MultiTenantConnectionProvider` and its delegate `TenantConnectionResolver`, we can't just say "register it when the user registers this or that bean".

Item 1 means we need to be able to tell if we're using database multitenancy vs. schema multitenancy. Though I suppose we could address this by assuming "if a datasource is configured, we're not using database multitenancy".

Item 2 means we need to be able to tell if we're using database/schema multitenancy vs. discriminator/no multitenancy, and for that I don't have a solution.

So... unless we significantly change our approach, I don't think we can get rid of `MultiTenancyStrategy`...

Alternative approaches to address item 2 include:

1.  Always enabling `MultiTenantConnectionProvider` and having it handle the single-tenancy case. This seems dodgy, as I suspect using a `MultiTenantConnectionProvider` has more effect than just changing the way connections are retrieved.
2. Not providing a default `TenantConnectionResolver`, or providing its implementation (code) but expecting it to be registered to CDI explicitly (e.g. through a `@Provider`). This would be poor developer experience IMO.
3. Exposing a setting to say basically "this datasource uses schema or datasource mutltitenancy" (as opposed to discriminator multitenancy or single tenancy).

None of these solutions seem particularly better than just exposing `quarkus.hibernate-orm.multitenancy`, IMO.

**Comment by lucamolteni** [comment]
@Sanne can you take a look at Yoann's answer so that we decide what to do with this PR?

**Comment by Sanne** [comment]
I agree with Yoann, it's not as simple to remove this. I also have to admit I don't remember all details though - someone would need to have a deeper look and see if we should just remove that "deprecated" comment or perform a more substantial cleanup.

**Comment by yrodiere** [comment]
IMO the long term strategy, which would incidentally allow us to remove `MultiTenancyStrategy`, would be to move all schema-based and database-based multitenancy to the datasource layer: #11861
It just makes sense, but it's work.

**Comment by Sanne** [comment]
> IMO the long term strategy, which would incidentally allow us to remove `MultiTenancyStrategy`, would be to move all schema-based and database-based multitenancy to the datasource layer: [#11861](https://github.com/quarkusio/quarkus/issues/11861) It just makes sense, but it's work.

I agree that it would be nice, but bear in mind that some features of ORM expose schema names / catalogue names. So having the schema fully "hidden" and managed by the lower layer might not always be an option.

**Comment by yrodiere** [comment]
> > IMO the long term strategy, which would incidentally allow us to remove `MultiTenancyStrategy`, would be to move all schema-based and database-based multitenancy to the datasource layer: [#11861](https://github.com/quarkusio/quarkus/issues/11861) It just makes sense, but it's work.
> 
> I agree that it would be nice, but bear in mind that some features of ORM expose schema names / catalogue names. So having the schema fully "hidden" and managed by the lower layer might not always be an option.

That's true, but we're talking about the datasource defining the default schema here, not enforcing it. Just like we do in Hibernate ORM. So we'd have the exact same limitations.

**Comment by Sanne** [comment]
Ok. Bear in mind that there might be efficiency implications. If the datasource has to "set" the current schema on each connection request, that's going to be a full roundtrip on some implementations. I hope I'm wrong on this, just making a note here that this is something that should be checked before committing to any particular design.

**Comment by yrodiere** [comment]
> If the datasource has to "set" the current schema on each connection request, that's going to be a full roundtrip on some implementations

I have bad news about the current implementation in Hibernate/Quarkus then 🙈 

https://github.com/quarkusio/quarkus/blob/5f4ff6f08dd10e632aa568f60f1e0382a2dd3819/extensions/hibernate-orm/runtime/src/main/java/io/quarkus/hibernate/orm/runtime/tenant/DataSourceTenantConnectionResolver.java#L78-L95

**Comment by Sanne** [comment]
right, that's what I'm referring to. We need to keep the cost of such approaches in mind.
